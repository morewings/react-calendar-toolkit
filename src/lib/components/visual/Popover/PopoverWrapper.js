import React, {useEffect} from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import {useTheme} from 'lib/features/theme';
import useOnClickOutside from 'lib/utils/useOnClickOutside';
import classes from './PopoverWrapper.module.css';

const PopoverWrapper = ({position, children, toggleDatepicker}) => {
  /* Apply postcss theme to containing div */
  const {ref, setRef, style} = useTheme();

  /* Close popover on click outside */
  useOnClickOutside(ref, () => {
    toggleDatepicker(false);
  });

  /* Close popover when user is scrolling page */
  useEffect(() => {
    const listener = () => {
      toggleDatepicker(false);
    };
    document.addEventListener('scroll', listener);
    return () => {
      document.removeEventListener('scroll', listener);
    };
  }, [toggleDatepicker]);

  return (
    <div
      style={style}
      ref={setRef}
      className={classnames({
        [classes.wrapper]: true,
        /* Conditional class to position Popover relative to Input */
        [classes[position]]: true,
      })}>
      <div
        className={classnames({
          [classes.triangle]: true,
          /* Conditional class to position triangle attachment relative to Input */
          [classes[position]]: true,
        })}
      />
      {children}
    </div>
  );
};

PopoverWrapper.propTypes = {
  children: PropTypes.node.isRequired,
  /** Method to toggle Datepicker visibility */
  toggleDatepicker: PropTypes.func.isRequired,
  /** Calculated position of Popover to fit into viewport */
  position: PropTypes.oneOf(['bottom', 'top', 'left', 'right']),
};

PopoverWrapper.defaultProps = {
  position: 'bottom',
};

export default PopoverWrapper;
