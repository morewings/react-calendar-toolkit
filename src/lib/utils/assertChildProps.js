import React, {Fragment} from 'react';

export const trackProps = jest.fn();

export const DummyComponent = props => {
  trackProps(props);
  return <Fragment>Child</Fragment>;
};
