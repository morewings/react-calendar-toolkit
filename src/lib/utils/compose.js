const compose = (...fns) => (...args) =>
  fns.reduceRight(
    (params, f) => (Array.isArray(params) ? f(...params) : f(params)),
    args
  );

export default compose;
