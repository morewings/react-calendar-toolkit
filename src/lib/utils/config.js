export default Object.freeze({
  supportedPrecisions: Object.freeze([
    'year',
    'month',
    'day',
    'hour',
    'minute',
  ]),
});
