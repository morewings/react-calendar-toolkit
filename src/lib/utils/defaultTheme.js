const defaultTheme = {
  gray: 'rgba(0, 0, 0, 0.38)',
  baseColor: '#43a047',
  borderRadius: '4px',
  /** Set font for datepicker */
  fontPrimary: 'Helvetica Neue',
  /** Font, which is applied, if --fontPrimary is not available */
  fontSecondary: 'Arial',
  /** Fallback font if two previous are unavailable */
  fontFallback: 'sans-serif',
  fontSize: '14px',
  fontWeight: '400',
  /** Applied to the title */
  fontSizeSmall: '10px',
  /** Applied to the header text */
  fontSizeBig: '24px',
  /** Set inner width */
  calendarWidth: '240px',
  /** Calculate total width */
  datePickerWidth: 'calc(var(--calendarWidth) + var(--innerPadding) * 2)',
  selectorHeight: '36px',
  calendarHeight: '200px',
  /** Set padding between elements */
  innerPadding: '12px',
  /** Day, month, year entries height */
  entryHeight: '32px',
  entryMargin: '1px',
  datepickerBorderRadius: 'var(--borderRadius)',
  entryBorderRadius: 'var(--borderRadius)',
  /** Background color of Calendar */
  bgColor: 'white',
  /** Base color for text */
  textColor: 'black',
  headerBgColor: 'var(--baseColor)',
  weekendTextColor: 'red',
  headerTextColor: 'white',
  borderColor: 'var(--gray)',
  modalBgColor: 'var(--gray)',
  weekDayColor: 'var(--gray)',
  entryHoverColor: 'rgba(0, 0, 0, 0.08)',
  entrySelectedColor: 'var(--baseColor)',
  entryHighlightColor: 'lightblue',
  inputWidth: '90px',
  inputPadding: '8px',
  inputBorderRadius: 'var(--borderRadius)',
  inputBorderColor: 'var(--gray)',
  inputHighlightColor: 'lightblue',
};

export default defaultTheme;
