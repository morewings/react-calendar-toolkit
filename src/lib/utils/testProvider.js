import React from 'react';
import PropTypes from 'prop-types';
import {Provider as LocaleProvider} from 'lib/features/locale/withLocaleContext';
import {Provider as ThemeProvider} from 'lib/features/theme/withTheme';
import {Provider as ModalProvider} from 'lib/features/modal/withModalProvider';
import {Provider as DatepickerProvider} from 'lib/features/datepicker/withDatepickerProvider';

export const DatepickerMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <DatepickerProvider>{props.children}</DatepickerProvider>
    </ThemeProvider>
  </LocaleProvider>
);

export const InputMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <ModalProvider>
        <DatepickerProvider>{props.children}</DatepickerProvider>
      </ModalProvider>
    </ThemeProvider>
  </LocaleProvider>
);

DatepickerMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

InputMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
