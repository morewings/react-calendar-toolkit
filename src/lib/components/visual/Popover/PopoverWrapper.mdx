---
name: PopoverWrapper
menu: Customize UI
---
import { Playground, Props } from 'docz';
import PopoverWrapper from './PopoverWrapper';

`PopoverWrapper` used to attach a triangle and position `Datepicker` relative to `Input`.

## Props

<Props of={PopoverWrapper} />



