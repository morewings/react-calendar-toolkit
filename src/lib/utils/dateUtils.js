import {
  formatWithOptions,
  isSameMonth,
  isSameYear,
  toDate,
  getTime,
  addMonths,
  addDays,
  addYears,
  isWithinInterval,
  isWeekend,
  isSameDay,
  isSameHour,
  getDate,
  startOfMonth,
  endOfMonth,
} from 'date-fns/fp';
import compose from 'lib/utils/compose';

/**
 * Unix timestamp
 * @typedef {number} UnixTimestamp
 */

/**
 * Date union type may be Date or Unix timestamp
 * @typedef {(Date|UnixTimestamp)} DateUnion
 */

/**
 * @function
 * @name formatDateWithLocale
 * @description Returns formatted date
 * @param {Object} options - Options object with locale
 * @param {string} pattern - Formatting pattern
 * @param {Date} date - Date object to apply format
 * @return {string} Formatted date
 */
export const formatDateWithLocale = formatWithOptions;

/**
 * @function
 * @name checkIsSameMonth
 * @description Checks if two dates belong to the same month
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameMonth = isSameMonth;

/**
 * @function
 * @name checkIsSameYear
 * @description Checks if two dates belong to the same year
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameYear = isSameYear;

/**
 * @function
 * @name checkIsSameDay
 * @description Checks if two dates belong to the same day
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameDay = isSameDay;

/**
 * @function
 * @name checkIsSameHour
 * @description Checks if two dates belong to the same hour
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameHour = isSameHour;

/**
 * @function
 * @name convertToTimestamp
 * @description Converts provided date to Unix timestamp
 * @param {Date} date - Date to convert
 * @return {UnixTimestamp} Unix timestamp
 */
export const convertToTimestamp = date => getTime(date);

/**
 * @function
 * @name convertToDate
 * @description Converts provided Unix timestamp to Date object
 * @param {number} timestamp - Date to convert
 * @return {Date}
 */
export const convertToDate = timestamp => toDate(timestamp);

/**
 * @function
 * @name convertToDayOfMonth
 * @description Converts provided Unix timestamp or Date to day of month number
 * @param {DateUnion} timestamp - Timestamp or Date to convert
 * @return {number} Day of the month
 */
export const convertToDayOfMonth = timestamp => getDate(timestamp);

/**
 * @function
 * @name checkIsWithinInterval
 * @description Checks if date belongs to the interval
 * @typedef {Object} Interval
 * @property {Date} start - Start of interval
 * @property {Date} end - End of interval
 * @param {DateUnion} date - Date to check
 * @param {Interval} interval - Time range to check
 * @return {boolean}
 */
export const checkIsWithinInterval = isWithinInterval;

/**
 * @function
 * @name checkIsWeekend
 * @description Checks if date belongs to the weekend
 * @param {DateUnion} date - Date to check
 * @return {boolean}
 */
export const checkIsWeekend = date => isWeekend(date);

/**
 * @function
 * @name matchDatesWithPrecision
 * @description Checks if dates are the same with given precision, eg belong to same day|month|year
 * @param {string} precision - Date to check
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const matchDatesWithPrecision = (precision, dateLeft, dateRight) =>
  ({
    day: checkIsSameDay,
    month: checkIsSameMonth,
    year: checkIsSameYear,
  }[precision](dateLeft, dateRight));

/**
 * @function
 * @name ceilMonth
 * @description Gets end of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const ceilMonth = endOfMonth;

/**
 * @function
 * @name floorMonth
 * @description Gets start of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const floorMonth = startOfMonth;

/**
 * @function
 * @name addMonth
 * @description Adds months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to add
 * @return {UnixTimestamp}
 */
export const addMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(amount, date);

/**
 * @function
 * @name subMonth
 * @description Subtracts months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to subtract
 * @return {UnixTimestamp}
 */
export const subMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(-amount, date);

/**
 * @function
 * @name addDay
 * @description Adds days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to add
 * @return {UnixTimestamp}
 */
export const addDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(amount, date);

/**
 * @function
 * @name subDay
 * @description Subtracts days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(-amount, date);

/**
 * @function
 * @name addYear
 * @description Adds years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of years to add
 * @return {UnixTimestamp}
 */
export const addYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(amount, date);

/**
 * @function
 * @name subYear
 * @description Subtracts years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(-amount, date);
