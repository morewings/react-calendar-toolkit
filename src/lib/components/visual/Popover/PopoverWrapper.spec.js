import React from 'react';
import {InputMockProvider as Provider} from 'lib/utils/testProvider';
import {render, fireEvent, act} from '@testing-library/react';
import PopoverWrapper from './PopoverWrapper';

const toggleDatepicker = jest.fn();
const Mock = () => <div data-testid="modalContent">Mock</div>;
const wrapper = document.createElement('div');
wrapper.setAttribute('id', 'wrapper');

const sibling = document.createElement('div');
sibling.setAttribute('id', 'sibling');
document.body.appendChild(sibling);

describe('PopoverWrapper', () => {
  beforeEach(() => {
    toggleDatepicker.mockClear();
  });

  it.each([['bottom'], ['top'], ['left'], ['right']])('renders', position => {
    const {container} = render(
      <PopoverWrapper position={position} toggleDatepicker={toggleDatepicker}>
        <Mock />
      </PopoverWrapper>,
      {
        container: document.body.appendChild(wrapper),
        wrapper: ({children}) => <Provider>{children}</Provider>,
      }
    );
    expect(container).toMatchSnapshot();
  });

  it('calls toggleDatepicker when clicked outside', () => {
    render(
      <PopoverWrapper position="bottom" toggleDatepicker={toggleDatepicker}>
        <Mock />
      </PopoverWrapper>,
      {
        container: document.body.appendChild(wrapper),
        wrapper: ({children}) => <Provider>{children}</Provider>,
      }
    );
    fireEvent.mouseDown(document.getElementById('sibling'));
    expect(toggleDatepicker).toHaveBeenCalledTimes(1);
    expect(toggleDatepicker).toHaveBeenCalledWith(false);
    fireEvent.touchStart(document.getElementById('sibling'));
    expect(toggleDatepicker).toHaveBeenCalledTimes(2);
    expect(toggleDatepicker).toHaveBeenLastCalledWith(false);
  });

  it('does not call toggleDatepicker when clicked inside', () => {
    const {getByTestId} = render(
      <PopoverWrapper position="bottom" toggleDatepicker={toggleDatepicker}>
        <Mock />
      </PopoverWrapper>,
      {
        container: document.body.appendChild(wrapper),
        wrapper: ({children}) => <Provider>{children}</Provider>,
      }
    );
    fireEvent.mouseDown(getByTestId('modalContent'));
    expect(toggleDatepicker).not.toHaveBeenCalled();
    fireEvent.touchStart(getByTestId('modalContent'));
    expect(toggleDatepicker).not.toHaveBeenCalled();
  });

  it('calls toggleDatepicker when scrolling', () => {
    render(
      <PopoverWrapper position="bottom" toggleDatepicker={toggleDatepicker}>
        <Mock />
      </PopoverWrapper>,
      {
        container: document.body.appendChild(wrapper),
        wrapper: ({children}) => <Provider>{children}</Provider>,
      }
    );
    act(() => {
      fireEvent.scroll(document);
    });

    expect(toggleDatepicker).toHaveBeenCalledTimes(1);
    expect(toggleDatepicker).toHaveBeenCalledWith(false);
  });
});
