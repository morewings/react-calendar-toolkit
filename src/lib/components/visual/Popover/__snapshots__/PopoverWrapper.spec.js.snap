// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopoverWrapper renders 1`] = `
<div
  id="wrapper"
>
  <div
    class="wrapper bottom"
    style="--gray: rgba(0, 0, 0, 0.38); --baseColor: #43a047; --borderRadius: 4px; --fontPrimary: Helvetica Neue; --fontSecondary: Arial; --fontFallback: sans-serif; --fontSize: 14px; --fontWeight: 400; --fontSizeSmall: 10px; --fontSizeBig: 24px; --calendarWidth: 240px; --datePickerWidth: calc(var(--calendarWidth) + var(--innerPadding) * 2); --selectorHeight: 36px; --calendarHeight: 200px; --innerPadding: 12px; --entryHeight: 32px; --entryMargin: 1px; --datepickerBorderRadius: var(--borderRadius); --entryBorderRadius: var(--borderRadius); --bgColor: white; --textColor: black; --headerBgColor: var(--baseColor); --weekendTextColor: red; --headerTextColor: white; --borderColor: var(--gray); --modalBgColor: var(--gray); --weekDayColor: var(--gray); --entryHoverColor: rgba(0, 0, 0, 0.08); --entrySelectedColor: var(--baseColor); --entryHighlightColor: lightblue; --inputWidth: 90px; --inputPadding: 8px; --inputBorderRadius: var(--borderRadius); --inputBorderColor: var(--gray); --inputHighlightColor: lightblue;"
  >
    <div
      class="triangle bottom"
    />
    <div
      data-testid="modalContent"
    >
      Mock
    </div>
  </div>
</div>
`;

exports[`PopoverWrapper renders 2`] = `
<div
  id="wrapper"
>
  <div
    class="wrapper top"
    style="--gray: rgba(0, 0, 0, 0.38); --baseColor: #43a047; --borderRadius: 4px; --fontPrimary: Helvetica Neue; --fontSecondary: Arial; --fontFallback: sans-serif; --fontSize: 14px; --fontWeight: 400; --fontSizeSmall: 10px; --fontSizeBig: 24px; --calendarWidth: 240px; --datePickerWidth: calc(var(--calendarWidth) + var(--innerPadding) * 2); --selectorHeight: 36px; --calendarHeight: 200px; --innerPadding: 12px; --entryHeight: 32px; --entryMargin: 1px; --datepickerBorderRadius: var(--borderRadius); --entryBorderRadius: var(--borderRadius); --bgColor: white; --textColor: black; --headerBgColor: var(--baseColor); --weekendTextColor: red; --headerTextColor: white; --borderColor: var(--gray); --modalBgColor: var(--gray); --weekDayColor: var(--gray); --entryHoverColor: rgba(0, 0, 0, 0.08); --entrySelectedColor: var(--baseColor); --entryHighlightColor: lightblue; --inputWidth: 90px; --inputPadding: 8px; --inputBorderRadius: var(--borderRadius); --inputBorderColor: var(--gray); --inputHighlightColor: lightblue;"
  >
    <div
      class="triangle top"
    />
    <div
      data-testid="modalContent"
    >
      Mock
    </div>
  </div>
</div>
`;

exports[`PopoverWrapper renders 3`] = `
<div
  id="wrapper"
>
  <div
    class="wrapper left"
    style="--gray: rgba(0, 0, 0, 0.38); --baseColor: #43a047; --borderRadius: 4px; --fontPrimary: Helvetica Neue; --fontSecondary: Arial; --fontFallback: sans-serif; --fontSize: 14px; --fontWeight: 400; --fontSizeSmall: 10px; --fontSizeBig: 24px; --calendarWidth: 240px; --datePickerWidth: calc(var(--calendarWidth) + var(--innerPadding) * 2); --selectorHeight: 36px; --calendarHeight: 200px; --innerPadding: 12px; --entryHeight: 32px; --entryMargin: 1px; --datepickerBorderRadius: var(--borderRadius); --entryBorderRadius: var(--borderRadius); --bgColor: white; --textColor: black; --headerBgColor: var(--baseColor); --weekendTextColor: red; --headerTextColor: white; --borderColor: var(--gray); --modalBgColor: var(--gray); --weekDayColor: var(--gray); --entryHoverColor: rgba(0, 0, 0, 0.08); --entrySelectedColor: var(--baseColor); --entryHighlightColor: lightblue; --inputWidth: 90px; --inputPadding: 8px; --inputBorderRadius: var(--borderRadius); --inputBorderColor: var(--gray); --inputHighlightColor: lightblue;"
  >
    <div
      class="triangle left"
    />
    <div
      data-testid="modalContent"
    >
      Mock
    </div>
  </div>
</div>
`;

exports[`PopoverWrapper renders 4`] = `
<div
  id="wrapper"
>
  <div
    class="wrapper right"
    style="--gray: rgba(0, 0, 0, 0.38); --baseColor: #43a047; --borderRadius: 4px; --fontPrimary: Helvetica Neue; --fontSecondary: Arial; --fontFallback: sans-serif; --fontSize: 14px; --fontWeight: 400; --fontSizeSmall: 10px; --fontSizeBig: 24px; --calendarWidth: 240px; --datePickerWidth: calc(var(--calendarWidth) + var(--innerPadding) * 2); --selectorHeight: 36px; --calendarHeight: 200px; --innerPadding: 12px; --entryHeight: 32px; --entryMargin: 1px; --datepickerBorderRadius: var(--borderRadius); --entryBorderRadius: var(--borderRadius); --bgColor: white; --textColor: black; --headerBgColor: var(--baseColor); --weekendTextColor: red; --headerTextColor: white; --borderColor: var(--gray); --modalBgColor: var(--gray); --weekDayColor: var(--gray); --entryHoverColor: rgba(0, 0, 0, 0.08); --entrySelectedColor: var(--baseColor); --entryHighlightColor: lightblue; --inputWidth: 90px; --inputPadding: 8px; --inputBorderRadius: var(--borderRadius); --inputBorderColor: var(--gray); --inputHighlightColor: lightblue;"
  >
    <div
      class="triangle right"
    />
    <div
      data-testid="modalContent"
    >
      Mock
    </div>
  </div>
</div>
`;
